### PR TITLE
Update Icons fix and Cloaking Device rework

### PR DESCRIPTION
**Pull request by NanakoAC**
I started today by firing up the profiler and spawning a couple hundred naked humans to test with, hoping to find something significant to fix for performance gains. This experiment was quite a success, as i found that human/update_icons was using a disproportionate amount of CPU time.

On investigation i discovered three issues:

1. A pretty significant bug where all dead or SSD human mobs were attempting to update their tail every life proc (even if they dont have one) which resulted in a call to update icons too. I've optimised this so that update icons will only be called if the tail state is changed, and the tail code won't even run for species which don't have tails.

2. Cloaking device code stuck in where it shouldn't be. Hence i decided this was a good opportunity to rework them. I moved their code out into their own file. I removed the special case code from update icons, and instead ported them to the modifier system i wrote a couple months back, its designed for just this sort of thing. Also rebalanced them a little: Cloaking devices now have limited power, roughly 5 minutes of sustained use (this works out to a lot more actual usage if you use it wisely), Cell can be removed/recharged/replaced added the cloaking sound effect, and added mouse_opacity 0 for cloaked mobs. This results in them being un-rightclickable and not seeable in the altclick menu either, making the cloaking generally more robust for hiding. Cloaking devices also added to traitor uplink. 14 TC cost, pretty expensive

3. The optimisation for lying states was written but not implemented for some reason. I fixed it so the icon won't be transformed unless it needs to be: When someone changes between lying/standing, or if they have a nonstandard scaling factor. Scaling factor is fully supported, although currently unused. I intend to make use of it in upcoming genetics work.

With these changes, update_icons is now mostly a nonissue, and thus a great deal of load caused by SSD players is cleaned up

In addition, i've reworked the Dropped proc due to the problems its caused, and lack of valid use cases for its more esoteric behaviour.

Moving an item between your hands, or placing it into a container, now uses a custom proc to remove the item instead of the default unequip one. Which does most of the same stuff but does not call Dropped. 

As a result, dropped now only fires in three situations
1. When an item is actually dropped
2. When an item is thrown
3. When an item is placed into a container:
Note that before this change, placing an object into a container called dropped twice. Once on the floor, and then once in the container. My change removes the first one. The second dropped call still happens, and it happens AFTER the object is inside the container, allowing it to check its location

In short, dropped has been redefined from its old usage of "left an inventory slot" to now mean "left a mob's direct contents" which is a far more useful definition programatically.

### CODE REVIEW

File: code/game/objects/items/weapons/cloaking_device.dm [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+	//Picked up or switched hands or worn
+	register_owner(user)
+
+
+/obj/item/weapon/cloaking_device/dropped(var/mob/user)
+	..()
+	spawn(1)//Things are dropped on the floor briefly when being put into containers or swapping hands
+	//Its dumb. This little hack works around it
+		var/mob/M = get_holding_mob()
+		if(!M)
```
Comment (skull132) [inline code/game/objects/items/weapons/cloaking_device.dm]: This makes no sense. If no M, register M. Why are you not using pickedUp or whatever the proc is...?

File: code/game/objects/items/weapons/cloaking_device.dm [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+	if (modifier)
+		modifier.stop(1)
+		modifier = null
+
+
+/obj/item/weapon/cloaking_device/attackby(obj/item/weapon/W as obj, mob/user as mob)
+	if(istype(W, /obj/item/weapon/cell))
+		if(!cell)
+			user.drop_item()
+			W.loc = src
```
Comment (skull132) [inline code/game/objects/items/weapons/cloaking_device.dm]: No forcemove.

File: code/game/objects/items/weapons/cloaking_device.dm [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+			cell = W
+			user << "<span class='notice'>You install a cell in [src].</span>"
+			update_icon()
+		else
+			user << "<span class='notice'>[src] already has a cell.</span>"
+
+	else if(istype(W, /obj/item/weapon/screwdriver))
+		if(cell)
+			cell.update_icon()
+			cell.loc = get_turf(src.loc)
```
Comment (skull132) [inline code/game/objects/items/weapons/cloaking_device.dm]: Forcemove.

File: code/game/objects/items/weapons/cloaking_device.dm [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+/obj/item/weapon/cloaking_device/process()
+	if (!cell || !cell.checked_use(power_usage*CELLRATE))
+		deactivate()
+		return
+
+
+/*
+	Modifier
+*/
+/datum/modifier/cloaking_device/activate()
```
Comment (skull132) [inline code/game/objects/items/weapons/cloaking_device.dm]: Assuming stop() qdels the modifier here.

File: code/game/objects/items/weapons/cloaking_device.dm (hunk 1 above)
Comment (skull132) [inline code/game/objects/items/weapons/cloaking_device.dm]: Also, for safety's sake. This should unregister the owner. Otherwise, funky shit can happen.

Comment (NanakoAC) [inline code/game/objects/items/weapons/cloaking_device.dm]: no it makes perfect sense
You see, if there's no M, then M is null. The register function will clear things as normal and set the owner to the passed null value. which is exactly what's desired when its on the floor

File: code/game/objects/items/weapons/cloaking_device.dm (hunk 4 above)
Comment (NanakoAC) [inline code/game/objects/items/weapons/cloaking_device.dm]: yes the stop proc handles everything

File: code/game/objects/items/weapons/cloaking_device.dm (hunk 1 above)
Comment (NanakoAC) [inline code/game/objects/items/weapons/cloaking_device.dm]: and i'm using the dropped proc because this is FOR when its dropped.

This code handles cases where the cloaking device is dropped on the floor or thrown, the register function will remove the stealth from the person who was holding it

Comment (skull132) [inline code/game/objects/items/weapons/cloaking_device.dm]: Okay. It's still stupid.

```
/obj/item/weapon/cloaking_device/dropped(var/mob/user)
    ..()
    register_owner(null)
```

And then whatever the heck the pickup proc is:
```
/obj/item/weapon/cloaking_device/onPikcup(var/mob/user)
    ..()
    register_owner(user)
```

Comment (NanakoAC) [inline code/game/objects/items/weapons/cloaking_device.dm]: good point actually lemme just revert that thing i did

**Comment by NanakoAC** [comment]
ive audited several functions and traced their code path. There is no special proc 

http://pastebin.com/c3Hv8NSM

It's either dropped or forceMove that has to be hooked into 
Also in this commit, i rearranged some inventory related code. Removed several functions from carbon.dm and human.dm and moved them into the nearest inventory.dm, as that's the place for most functions involving manipulating obj/items

File: code/modules/mob/inventory.dm [hunk 5]
```diff
@@ -245,3 +245,60 @@ var/list/slot_equipment_priority = list( \
...
+
+	if(!item) return //Grab processing has a chance of returning null
+
+
+	src.remove_from_mob(item)
+	item.loc = src.loc
+
+	//actually throw it!
+	if (item)
+		src.visible_message("\red [src] has thrown [item].")
```
Comment (Lohikar) [inline code/modules/mob/inventory.dm]: Spans.

Comment (NanakoAC) [inline code/modules/mob/inventory.dm]: i didnt create these, just moved them between files

Comment (Lohikar) [inline code/modules/mob/inventory.dm]: Doesn't mean you shouldn't fix them.
